sound/soc/sof/sof-audio.c

```diff
@@ -350,7 +348,7 @@ sof_prepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget
 	if (!widget_ops)
 		goto sink_prepare;
 
-	if (!widget_ops[widget->id].ipc_prepare || swidget->prepared)
+	if (!widget_ops[widget->id].ipc_prepare && !swidget->use_count)
```

Comment:
This works as long as we have always prepare followed by setup (where the reference count is increased).

If this is not the case and we can have multiple calls to prepare, then this would not work.

In the first case, there should be a deeper explanation in the commit message as to why this is ok.

Reply:
@plbossart @ranj063 On the second thought, I think it is better to keep the prepared flag. Think about if a widget is prepared, but fail to setup. The use_count will be 0, but it is prepared. It will likely be a problem.

Reply:
But @bardliao don't we call unprepare if widget setup fails?

Reply:
> But @bardliao don't we call unprepare if widget setup fails?

Yes, we call sof_walk_widgets_in_order SOF_WIDGET_UNPREPARE if sof_walk_widgets_in_order SOF_WIDGET_SETUP failed. But what if one of the widget is failed and others are success? Those widgets that have already set up will not be unprepared since their use_count is 1. My point is that prepare and setup are handled in different helper functions, someone can easily break it even if it works well now.

Reply:
@bardliao when widget setup fails, we unroll and free all widgets that were previously set up in sof_set_up_widgets_in_path(). I am fairly certain the prepare flag is redundant